### PR TITLE
Design Doc: Session

### PR DESCRIPTION
**Pull request by helinwang**
[Here](https://github.com/helinwang/Paddle/blob/session/doc/design/refactor/session.md) is better for review.

Fixes: https://github.com/PaddlePaddle/Paddle/issues/4552

### CODE REVIEW

File: doc/design/session.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# Design Doc: Session
+
+## Abstract
+
+This design doc proposes to have an object called *Session* which
```
Comment (typhoonzero) [inline doc/design/session.md]: Is these design came out from https://github.com/PaddlePaddle/Paddle/pull/3811#issuecomment-328180377?

If so, you probably need to add descriptions like "session is able to distinguish running a graph locally or remotely, using CPU or GPU, using one device or more"

File: doc/design/session.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+encapsulates the environment in which the computation graph is
+executed.
+
+## Background
+
+A computation graph is executed in an environment which contains the
+[scope](./scope.md) and other states. PaddlePaddle used to only have
```
Comment (typhoonzero) [inline doc/design/session.md]: Sorry, what do you mean by "other states"?

File: doc/design/session.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+Session has two methods: `eval` and `close`. `eval` executes the
+target OP in a given graph, and `close` closes the session and
+releases all related resources:
+
+```Python
+a = paddle.constant(1.0)
+b = paddle.constant(2.0)
+c = a + b
+sess = paddle.session()
```
Comment (typhoonzero) [inline doc/design/session.md]: Some thoughts of mine:

- An NN training job must contain: graph(containing sub nets like init_net, forward_net, backward_net, opt_net); scope containing tensors as parameters; hyper parameters(learning_rate, batch_size, etc.); settings(cluster or not, devices, quotas, node ip address etc.), so here may be something like:
      sess = paddle.session(graph, scope_list, settings)
      # or
      sess = paddle.remote_session(graph, scop_list, cluster_settings)

- states can be stored all in some "scope" by creating scope for storing tensor for forward and backward,  and for storing hyper parameters changing states.

File: doc/design/session.md (hunk 2 above)
Comment (Superjom) [inline doc/design/session.md]: need to specify one session contains one scope?

File: doc/design/session.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+## Abstract
+
+This design doc proposes to have an object called *Session* which
+encapsulates the environment in which the computation graph is
+executed.
+
+## Background
+
+A computation graph is executed in an environment which contains the
```
Comment (wangkuiyi) [inline doc/design/session.md]: Try not to use passive voice.

> A computation graph runs in an environment

File: doc/design/session.md (hunk 2 above)
Comment (wangkuiyi) [inline doc/design/session.md]: "used to only have" => "used only to have" or "used to have only". Indeed, here we are describing the current statues, so it should be

> The current design has an implicit session ...

File: doc/design/session.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+This has the limitation that the user can not create two independent
+environments. For example, in reinforcement learning, the user may
+want to have a stale model for inference and a fresh model for
+training, and only replace the stale model with the fresh model
+periodically. Also, we have no concept that can encapsulate a remote
+environment that could execute a computation graph.
+
+## Session
+
+Session is an object that owns all runtime states such as scope,
```
Comment (wangkuiyi) [inline doc/design/session.md]: Session is an object ==> A session is an object

File: doc/design/session.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+encapsulates the environment in which the computation graph is
+executed.
+
+## Background
+
+A computation graph is executed in an environment which contains the
+[scope](./scope.md) and other states. PaddlePaddle used to only have
+an implicit global session on which `paddle.eval()` is executed.
+
+This has the limitation that the user can not create two independent
```
Comment (wangkuiyi) [inline doc/design/session.md]: Here is a broken logic.  It claimed but didn't explain why users cannot have two environments. The second sentence is to claim that it is necessary to have two environments.

From the text above, readers would ask it seems that what we need is two Scope instances, but not a new class Session.

File: doc/design/session.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+be executed on the Paddle Cloud. In this way, the user can control a
+cluster from her local computer:
+
+```Python
+reader = paddle.reader.recordio("/pfs/home/peter/mnist-train-*") # data stored on Paddle Cloud
+image = reader.column(0)
+label = reader.column(1)
+fc1 = paddle.op.fc(image, size=256, act="sigmoid")
+fc2 = paddle.op.fc(fc1, size=10, act="softmax")
+cost = paddle.op.cross_entropy(fc2)
```
Comment (putcn) [inline doc/design/session.md]: should this line be `cost = paddle.op.cross_entropy(fc2, **label**)` ?

**Comment by QiJune** [comment]
It's better to give a C++ or Python class definition of Session.

File: doc/design/session.md (hunk 2 above)
Comment (jacquesqiao) [reply doc/design/session.md]: scope, device, context etc.

Comment (dzhwinter) [reply doc/design/session.md]: I have the same question with @typhoonzero , @jacquesqiao  do you mean the *Session* contains runtime resources? 
e.g. already allocated memory in scope, occupied device, etc?

File: doc/design/session.md (hunk 1 above)
Comment (dzhwinter) [inline doc/design/session.md]: Maybe the *Session* try to solve the problem of 

1. unify the runtime sources management between a local machine and distributed environment.
2. Replace the *global scope* with a named concept, which holds resources explicitly. 

is it correct ?

Comment (helinwang) [reply doc/design/session.md]: Great point! Added "The Session is able to distinguish running a graph locally or remotely, using CPU only or using one or more GPUs."

File: doc/design/session.md (hunk 2 above)
Comment (helinwang) [reply doc/design/session.md]: Sorry for the late reply. @dzhwinter The environment contains runtime resources. The session  is a "owner" of these runtime resources.

Comment (helinwang) [reply doc/design/session.md]: Sorry for the late reply. Added "This indicates different sessions have different scopes.".

File: doc/design/session.md (hunk 3 above)
Comment (helinwang) [reply doc/design/session.md]: Agree with
```
sess = paddle.session(graph, scope_list, settings)
# or
sess = paddle.remote_session(graph, scop_list, cluster_settings)
```

> states can be stored all in some "scope" by creating scope for storing tensor for forward and backward, and for storing hyper parameters changing states.

Yes the variable states are stored in "scope". One session means one scope (just added into the doc).

File: doc/design/session.md (hunk 4 above)
Comment (helinwang) [reply doc/design/session.md]: Sorry for the late reply! Done.

File: doc/design/session.md (hunk 2 above)
Comment (helinwang) [reply doc/design/session.md]: Thanks! Done.

File: doc/design/session.md (hunk 6 above)
Comment (helinwang) [reply doc/design/session.md]: Thanks! Explained why user cannot have two environments, and also changed wording so that reader will know we need a new class Session.

File: doc/design/session.md (hunk 5 above)
Comment (helinwang) [reply doc/design/session.md]: Thanks! Done.

File: doc/design/session.md (hunk 7 above)
Comment (helinwang) [reply doc/design/session.md]: Thanks! Done.

File: doc/design/session.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+inference and a fresh model for training, and only replace the stale
+model with the fresh model periodically. Also, we have no concept that
+can encapsulate a remote environment that could execute a computation
+graph.
+
+We need a session concept to address above issues.
+
+## Session
+
+A session is an object that owns all runtime states such as scope,
```
Comment (typhoonzero) [inline doc/design/session.md]: Should define what a session owns exactly, like:

- The "graph" to run locally or remotely.
- Exactly ***one*** scope, containing all tensor variables for all the subnets.
- Settings and hyper-parameters.

Comment (helinwang) [reply doc/design/session.md]: Thanks! Updated, please review.

File: doc/design/session.md (hunk 1 above)
Comment (helinwang) [reply doc/design/session.md]: Yes.
Updated, please review:)

**Comment by helinwang** [comment]
@QiJune 

> It's better to give a C++ or Python class definition of Session.

Added Python interface, please review.

File: doc/design/refactor/session.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+```
+eval(
+    targets,
+    feed_dict=None,
+)
+```
+
+Evaluates the target Operations or Variables in `targets`.
+
+- *targets*: the evaluation targets. Can be a single Operation or
```
Comment (typhoonzero) [inline doc/design/refactor/session.md]: Should targets be an instance of "Block"(graph)?

File: doc/design/refactor/session.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+  Variable, or a list with the Operations or Variables as elements.
+
+  The value returned by `eval()` has the same shape as the `target`
+  argument.
+
+  The computation graph is implicitly inferred from the targets.
+
+- *feed_dict*: a dictionary that contains the tensors which overrides
```
Comment (typhoonzero) [inline doc/design/refactor/session.md]: Is this the input data?

File: doc/design/refactor/session.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+  the edges of the computation graph.
+
+```
+close()
+```
+
+Closes the session. Calling this method releases the scope.
```
Comment (typhoonzero) [inline doc/design/refactor/session.md]: Need `save()` and `load()` also to do checkpointing.

File: doc/design/refactor/session.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+```
+session(
+    gpu_ids=None
+)
+```
+
+Creates a new session. One session owns one scope, so creating
+multiple sessions will create different scopes.
+
+- *gpu_ids*: a single `int` or a list of `int` of the GPU IDs to be
```
Comment (typhoonzero) [inline doc/design/refactor/session.md]: Setting up devices can take advantage of paddle v1 design, which default set `gpu_ids` to all of the available GPUs.

Can we use a string to specify devices, because there may be other devices than GPU, like FPGA, what TF does is `/job:worker/task:1/gpu:0` also can be `/job:worker/task:1/fpga:0`

File: doc/design/refactor/session.md (hunk 10 above)
Comment (helinwang) [reply doc/design/refactor/session.md]: Yes, not only the input data, can override any edge as well. E.g.,:
```
a = pd.constant(1.0, name="a")
b = pd.constant(2.0)
c = pd.mul(a,b)
sess.eval(targets=c, feed_dict={"a":3.0}) # returns 6.0
```

File: doc/design/refactor/session.md (hunk 11 above)
Comment (helinwang) [reply doc/design/refactor/session.md]: `save` and `load` will be OPs, so user will need to run something like `sess.eval(targets=save)` or `sess.eval(targets=load)`

File: doc/design/refactor/session.md (hunk 12 above)
Comment (helinwang) [reply doc/design/refactor/session.md]: Great idea! Added.

Comment (QiJune) [reply doc/design/refactor/session.md]: I thinking what's the difference between Session and [Executor](https://github.com/QiJune/Paddle/blob/9447858aa778bc4fe18419c746c2c6997d822633/doc/design/block_execution.md#run-time-analysis). The Executor has a Run interface to execute a ProgramDesc created in compile-time. DeviceContext(CPUDeviceContext/CUDADeviceContext) is created and managed by Executor.
And Maybe Executor is a data member of Session. Session will get a target and some device ids. The target is parsed to get a ProgramDesc. Then, the ProgramDesc and device ids are passed to Executor. Executor will created a DeviceContextManeger according to device ids. At last,  the ProgramDesc will be executed in specific hardwares.

File: doc/design/refactor/session.md (hunk 9 above)
Comment (helinwang) [reply doc/design/refactor/session.md]: Targets will be a OP the output of a OP (Var).
To make the relationship more clear, I have added:

  The PaddlePaddle program is represented by the [ProgramDesc](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/program.md), `eval` will infer the ProgramDesc from the given *targets* and run the PaddlePaddle program.

File: doc/design/refactor/session.md (hunk 12 above)
Comment (helinwang) [reply doc/design/refactor/session.md]: @QiJune Great! That is a clear logic. We could add one more step ProgramOptimizer (currently called Converter) between Session and Executor. Please see this graph for more detail: https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/refactor/distributed_architecture.md#local-training-architecture (In the graph, "PaddlePaddle runtime" means the Executor. Btw, there are many names of the same thing, we need to decide on the naming).

Comment (dzhwinter) [reply doc/design/refactor/session.md]: After the offline discussion with @helinwang , I believe `Session` is 
- a manager of the resources, which is a more high-level concept than DeviceContextManager. It owns the resources.
- The job handler.
   - it's will construct a new graph according to the `targets` user given. 
   - Interact with remote cluster or local machine, fetch/feed tensor, start/stop running a graph....
   - The close interface allows user release the resource it owns.
The `Executor` is a runtime concept, which has nothing to do with Session.

Comment (helinwang) [reply doc/design/refactor/session.md]: > it's will construct a new graph according to the targets user given.

The session will not "construct a new graph", it will send the graph to the converter, and the converter will construct a new graph.

**Comment by helinwang** [comment]
@dzhwinter 

> it's will construct a new graph according to the targets user given.

The session will not "construct a new graph", it will send the graph to the converter, and the converter will construct a new graph.

File: doc/design/refactor/session.md [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+  the edges of the computation graph.
+
+  feed_dict not only can provide the input data, it can override any
+  OP's input as well:
+
+  ```python
+  a = pd.constant(1.0, name="a")
+  b = pd.constant(2.0)
+  c = pd.mul(a,b)
+  sess.eval(targets=c, feed_dict={"a":3.0}) # returns 6.0
```
Comment (dzhwinter) [inline doc/design/refactor/session.md]: constant can be changed value looks weird. Maybe name them with `variable` is better?

File: doc/design/refactor/session.md [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+
+### Create a Local Session
+
+```python
+session(
+    devices=None
+)
+```
+
+Creates a new session. One session owns one scope, so creating
```
Comment (dzhwinter) [inline doc/design/refactor/session.md]: This statement is confusing. One session owns **at least** one scope, namely, global scope in one single session, right? 
Or you mean that one session will have exactly one scope?

File: doc/design/refactor/session.md [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+multiple sessions will create different scopes.
+
+- *devices*: a single `string` or a list of `string` of device names,
```
Comment (dzhwinter) [inline doc/design/refactor/session.md]: Maybe we really need a `SessionConfig` here?  I mean, if one session cannot fully utilize the GPU resource, then another session may also own the same GPU.

In my view, we need to submit the config for this round of session run call. If it is a local run call, we provide the local config,  vice versa, we provide the cluster config. Even more, if it is an inference run call, we may provide another config, which is totally different from those ones above.
We can leave these complexities to be solved in the future, but we need to figure the concept clear.

Comment (helinwang) [reply doc/design/refactor/session.md]: > if one session cannot fully utilize the GPU resource, then another session may also own the same GPU.

Devices only means the devices that the session uses, multiple sessions can use the same device. I have added "Multiple sessions can use the same device." to clear up this point.

> If it is a local run call, we provide the local config, vice versa, we provide the cluster config

Local session is created by `paddle.session`, remote session is created by `paddle.remote_session`.

> if it is an inference run call, we may provide another config

In my view, inferencing is just user specifying a inference target, which is no different than training (specifying a training target). They should use the same kind of session. The layers below session should do the optimization (e.g., based on batch size) transparent to which session is used.

File: doc/design/refactor/session.md (hunk 13 above)
Comment (helinwang) [reply doc/design/refactor/session.md]: Thanks! Done.

File: doc/design/refactor/session.md (hunk 14 above)
Comment (helinwang) [reply doc/design/refactor/session.md]: Yes, the global scope. Done.

File: doc/design/refactor/session.md (hunk 15 above)
Comment (dzhwinter) [reply doc/design/refactor/session.md]: Ok, I see.

**Comment by dzhwinter** [comment]
We had discussed this module for a long-term and had reached an agreement, so it's better to merge it. For anyone who has any questions or different view, we can have an offline meeting. :)
